diff --git a/js/src/builtin/ReflectParse.cpp b/js/src/builtin/ReflectParse.cpp
index c81e06fe9a3b..bbc4ac334971 100644
--- a/js/src/builtin/ReflectParse.cpp
+++ b/js/src/builtin/ReflectParse.cpp
@@ -1704,7 +1704,7 @@ bool NodeBuilder::classDefinition(bool expr, HandleValue name,
   ASTType type = expr ? AST_CLASS_EXPR : AST_CLASS_STMT;
   RootedValue cb(cx, callbacks[type]);
   if (!cb.isNull()) {
-    return callback(cb, name, heritage, block, pos, dst);
+    return callback(cb, name, opt(heritage), block, pos, dst);
   }
 
   return newNode(type, pos, "id", name, "superClass", heritage, "body", block,
